velero-plugins/pod/backup.go

```diff
@@ -0,0 +1,47 @@
+package pod
+
+import (
+	"encoding/json"
+
+	"github.com/sirupsen/logrus"
+	v1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+	"github.com/vmware-tanzu/velero/pkg/podvolume"
+	corev1API "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// BackupPlugin is a backup item action plugin for Heptio Ark.
+type BackupPlugin struct {
+	Log logrus.FieldLogger
+}
+
+// AppliesTo returns a velero.ResourceSelector that applies to pods.
+func (p *BackupPlugin) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"pods"},
+	}, nil
+}
+const buildPodVolumesToExclude = "buildworkdir,container-storage-root,build-blob-cache"
+// Execute copies local registry images into migration registry
+func (p *BackupPlugin) Execute(input runtime.Unstructured, backup *v1.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+	pod := corev1API.Pod{}
+	itemMarshal, _ := json.Marshal(input)
+	json.Unmarshal(itemMarshal, &pod)
+	p.Log.Infof("[pod-backup] pod: %s", pod.Name)
+	// if pod is a build pod and it might already be completed.
+	// we still want build pods to be in the backup so skip volumes to avoid restic backup failure of a completed build pod
+	if pod.Labels["openshift.io/build.name"] != "" || pod.Annotations["openshift.io/build.name"] != "" {
+		if pod.Annotations[podvolume.VolumesToExcludeAnnotation] == "" {
```

Comment:
nil check here too:
```
		if pod.Annotations == nil || pod.Annotations[podvolume.VolumesToExcludeAnnotation] == "" {
```

Reply:
I do have one question about the logic here. It looks like we're only adding `buildPodVolumesToExclude` if there are no volumes currently excluded. This means that if there are different build pod volumes (or a subset of the above-listed pod volumes) excluded already, we don't add anything here. Is that reasonable? I guess it's an edge case, but this means if a user sets a different set of exclusion volumes, we don't do anything. Is that what we want, or do we want to merge this list with the already-annotated list?

Reply:
I was under the impression that we don't want to mess with user volumesToExclude annotation which is typically not expected to be present on a build pod.

Reply:
If the user is already doing something here.. let them keep their config.